MT3-565 #ready-for-test Boolean checkbox for post visit actions

LBHELewis opened the pull request:
So I've created a Boolean checkbox, and I've created an example of it in the `id` step. My concern is that actually looking at the design this should be inside the `TextareaDetails` component, and I don't know if that's something we have the ability to achieve? (A DynamicComponent inside a DynamicComponent)

LBHELewis committed:
Boolean checkbox

erbridge commented on components/BooleanCheckbox.tsx:
```diff
@@ -0,0 +1,62 @@
+
+import PropTypes from "../helpers/PropTypes";
+
+export type BooleanCheckboxProps = DynamicComponentControlledProps<string> & {
```
```suggestion
export type BooleanCheckboxProps = DynamicComponentControlledProps<boolean> & {
```

This should be using a boolean value in the database. I think we want it to have a fixed value (e.g. `"enabled"`), and we check if the value is checked. This _might_ be easier if it didn't use `Checkboxes`.

erbridge reviewed (commented):
I think we might want to make `notes: string` values in the database be `notes: { value: string; isPostVisitAction: boolean; }`. That way it's a single component and we combine the textarea and checkbox in the wrapper.

LBHELewis commented:
@erbridge I've changed how this works now based on the comments above. Please could you take another look?

erbridge commented on components/TextAreaDetails.tsx:
```diff
@@ -29,13 +37,53 @@ export const TextAreaDetails: React.FunctionComponent<Props> = props => {
+            return onValueChange({
+              value: value.value,
+              isPostVisitAction: false
+            });
```
This could be simplified to:

```
            return onValueChange({
              value: value.value,
              isPostVisitAction: checkboxValue[0] === "true"
            });
```

erbridge replied:
Or even:

```
            return onValueChange({
              value: value.value,
              isPostVisitAction: checkboxValue.length > 0
            });
```

erbridge commented on components/TextAreaDetails.tsx:
```diff
@@ -29,13 +37,53 @@ export const TextAreaDetails: React.FunctionComponent<Props> = props => {
+  if (includeCheckbox) {
+    const checkbox = {
+      label: "Create a post-visit action",
+      value: true
```
```suggestion
      value: "true"
```

erbridge commented on steps/id-and-residency/id.tsx:
```diff
@@ -126,43 +125,29 @@ const step: ProcessStepDefinition<ProcessDatabaseSchema, "id"> = {
+            contentBeforeTextArea?: React.ReactNode;
+            contentAfterTextArea?: React.ReactNode;
+            includeCheckbox?: boolean;
           },
```
This type cast is getting unwieldy. Perhaps we could cast to an exported `TextAreaDetailsProps` interface instead.

erbridge commented on components/TextAreaDetails.tsx:
```diff
@@ -3,13 +3,20 @@ import {
-type Props = DynamicComponentControlledProps<string> & {
+type Props = DynamicComponentControlledProps<{
+  value: string;
+  isPostVisitAction?: boolean;
```
I think this should be required, and default to `false` when we're not showing a checkbox.

erbridge replied:
This would make the types in the schema required, too.

erbridge reviewed (commented):
Looks like a sensible approach. Appreciate not making this a generic component for the checkbox. Would you mind editing the title to be more descriptive?

LBHELewis commented:
@erbridge I've just updated this. I realised that the checkbox actually needed to be on the normal checkbox and not just TextAreaDetails, since it sometimes appears with the textbox without being inside a details component. Have made the relevant changes.

Gweaton commented on storage/ProcessDatabaseSchema.ts:
```diff
@@ -26,88 +26,136 @@ type ProcessDatabaseSchema = NamedSchema<
           canEnterAll: string;
-          notes: string;
+          notes: {
+            value: string;
```
I tried to run this locally and saw some errors in the console. After a discussion with @erbridge we realised that we need to add a migration for the new `notes` field on each property. 

I've been informed you're being pulled onto other projects @LBHELewis and may be short on time, so I've made a start on trying to fix this, but it's currently very much a [work in progress](https://github.com/LBHackney-IT/mat-process-thc/commit/a54ef9de655e7b72995944d4f285c79ec62ec607).

I'll keep going with it and clear up any issues, but just wanted to keep you posted!

LBHELewis replied:
@Gweaton Amazing, thank you George.

Gweaton committed:
Add migration to new Note structure

Notes are now an object with `value` and `isPostVisitAction` fields.

Bump db to version 5

Gweaton committed:
Add Lodash to project for CloneDeep

This will be used to check which keys have a "notes" field

Gweaton commented on storage/ProcessDatabaseSchema.ts:
```diff
@@ -190,20 +208,32 @@ type ProcessDatabaseSchema = NamedSchema<
+          value: string;
+          isPostVisitAction?: boolean;
+        };
+        befriendingNotes: { value: string; isPostVisitAction?: boolean };
```
I think I might need to add a migration for these as well, as they're not caught by the current migration that looks for just `notes`

erbridge replied:
Yes. Maybe specify an array of keys to migrate and iterate through them?

erbridge replied:
I'd give them the same type as the other notes values, too.

erbridge commented on components/TextAreaWithCheckbox.tsx:
```diff
@@ -0,0 +1,111 @@
+  };
+  name: string;
+  rows?: number;
+  includeCheckbox?: boolean;
```
Would this ever be `false`? I think we don't want this prop, and instead always add the checkbox (the same says "with checkbox", so having it be optional is a bit weird.

erbridge commented on steps/household/household.tsx:
```diff
@@ -15,7 +15,10 @@ import {
-import { TextAreaDetails } from "../../components/TextAreaDetails";
+import {
+  TextAreaDetails,
+  Props as TextAreaDetailsProps
```
Let's rename them to this in the source file instead of renaming them on import.

erbridge commented on steps/household/household.tsx:
```diff
@@ -138,15 +143,13 @@ const step = {
+            includeCheckbox: true
+          } as TextAreaDetailsProps,
+          defaultValue: { value: "", isPostVisitAction: false },
+          emptyValue: { value: "", isPostVisitAction: false },
```
Ah. I've realized that this doesn't work properly. `remultiform` checks for equality, which will never be true for objects, even if they look the same. For now, I think it's ok, but I think we probably need something in the backlog to support this properly in `remultiform`. That means also adding it to `ComponentValue` and reverting the change above to `ProcessStepDefinition`.

erbridge commented on storage/ProcessDatabaseSchema.ts:
```diff
@@ -168,7 +180,10 @@ type ProcessDatabaseSchema = NamedSchema<
+        notes: {
+          value: string;
+          isPostVisitAction?: boolean;
+        };
```
These need to be `Note`s, too.

erbridge commented on storage/ResidentDatabaseSchema.ts:
```diff
@@ -15,7 +15,10 @@ type ResidentDatabaseSchema = NamedSchema<
+        notes: {
+          value: string;
+          isPostVisitAction: boolean;
+        };
```
These could probably do with sharing the `Note` interface. That might mean pulling it out into a `DatabaseSchema.d.ts` file for schema types to avoid a circular reference.

erbridge replied:
We'll need a migration for these, too. Remember, they're in a different database from the others.

erbridge commented on storage/Storage.ts:
```diff
@@ -41,6 +44,28 @@ const migrateProcessData = async (
+
+      for (const value of Object.values(processData[storeName])) {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        if (value && (value as any).notes !== undefined) {
```
Let's flip this to a guard statement instead to reduce the arrow pattern here.
